gem/optimise.py

```diff
@@ -58,6 +61,70 @@ def ffc_rounding(expression, epsilon):
 
 
 @singledispatch
+def _replace_div(node, self):
+    """Replace division with multiplication
+
+    :param node: root of expression
+    :param self: function for recursive calls
+    """
+    raise AssertionError("cannot handle type %s" % type(node))
+
+
+_replace_div.register(Node)(reuse_if_untouched)
+
+
+@_replace_div.register(Division)
+def _replace_div_division(node, self):
+    a, b = node.children
+    if isinstance(b, Literal):
+        return Product(self(a), Literal(1.0/b.array))
+    else:
+        return Product(self(a), Division(Literal(1.0), self(b)))
+
+
+def replace_division(expressions):
+    """Replace divisions with multiplications in expressions"""
+    mapper = Memoizer(_replace_div)
+    return list(map(mapper, expressions))
+
+
+@singledispatch
+def _reassociate_product(node, self):
+    """Rearrange sequence of chain of products in increasing order of node rank.
+     For example, the product ::
+
+        a*b[i]*c[i][j]*d
+
+    are reordered as ::
+
+        a*d*b[i]*c[i][j]
+
+    :param node: root of expression
+    :return: reassociated product node
+    """
+    raise AssertionError("cannot handle type %s" % type(node))
+
+
+_reassociate_product.register(Node)(reuse_if_untouched)
+
+
+@_reassociate_product.register(Product)
+def _reassociate_product_prod(node, self):
+    # collect all factors of product, sort by rank
+    # should use more sophisticated method later on for optimal result
+    comp_func = lambda x: len(x.free_indices)
+    factors = sorted(collect_terms(node, Product), key=comp_func)
+    # need to optimise away iterator <==> list
+    new_factors = list(map(self, factors))  # recursion
+    return reduce(Product, new_factors)
```

Comment:
Good point! They probably won't be DAG'ed. Will need to refine this (also need to change sumproduct_2_node)